drivers/led/is31fl3216a.c

```diff
@@ -0,0 +1,151 @@
+#define IS31FL3216A_REG_CTL_2		0x02
+#define IS31FL3216A_REG_LIGHT_EFFECT	0x03
+#define IS31FL3216A_REG_CHANNEL_CONFIG	0x04
+#define IS31FL3216A_REG_PWM_FIRST	0x10
+#define IS31FL3216A_REG_PWM_LAST	0x1F
+#define IS31FL3216A_REG_UPDATE		0xB0
+
+#define IS31FL3216A_MAX_LEDS		16
+
+LOG_MODULE_REGISTER(is31fl3216a, CONFIG_LED_LOG_LEVEL);
+
+struct is31fl3216a_cfg {
+	struct i2c_dt_spec i2c;
+};
+
+static int is31fl3216a_write_reg(const struct i2c_dt_spec *i2c, uint8_t reg,
+				 uint8_t val)
+{
+	uint8_t buffer[2] = {reg, val};
+	int status;
+
+	status = i2c_write_dt(i2c, buffer, sizeof(buffer));
+	if (status) {
+		LOG_ERR("Could not write register 0x%02x: %i", reg, status);
+	}
+
+	return status;
+}
+
+static int is31fl3216a_led_set_brightness(const struct device *dev,
+					  uint32_t led, uint8_t value)
+{
+	const struct is31fl3216a_cfg *config = dev->config;
+	uint8_t pwm_reg = IS31FL3216A_REG_PWM_LAST - led;
+	int status;
+	uint8_t pwm_value;
+
+	if (led > IS31FL3216A_MAX_LEDS - 1) {
+		return -ENODEV;
+	}
+
+	if (value > 100) {
+		return -EINVAL;
+	}
+
+	pwm_value = (0xFFU * value) / 100;
+	status = is31fl3216a_write_reg(&config->i2c, pwm_reg, pwm_value);
+	if (status) {
+		return status;
+	}
+
+	return is31fl3216a_write_reg(&config->i2c, IS31FL3216A_REG_UPDATE, 0);
+}
+
+static int is31fl3216a_led_on(const struct device *dev, uint32_t led)
+{
+	return is31fl3216a_led_set_brightness(dev, led, 100);
+}
+
+static int is31fl3216a_led_off(const struct device *dev, uint32_t led)
+{
+	return is31fl3216a_led_set_brightness(dev, led, 0);
+}
+
+static int is31fl3216a_init_registers(const struct i2c_dt_spec *i2c)
+{
+	int status;
+
+	status = is31fl3216a_write_reg(i2c, IS31FL3216A_REG_CTL_1, 0xFF);
+	if (status) {
+		return status;
+	}
+
+	status = is31fl3216a_write_reg(i2c, IS31FL3216A_REG_CTL_2, 0xFF);
+	if (status) {
+		return status;
+	}
+
+	status = is31fl3216a_write_reg(i2c, IS31FL3216A_REG_LIGHT_EFFECT, 0x00);
+	if (status) {
+		return status;
+	}
+
+	status = is31fl3216a_write_reg(i2c, IS31FL3216A_REG_CHANNEL_CONFIG,
+				       0x00);
+	if (status) {
+		return status;
+	}
+
+	for (int i = IS31FL3216A_REG_PWM_FIRST;
+	     i <= IS31FL3216A_REG_PWM_LAST;
+	     i++) {
+		status = is31fl3216a_write_reg(i2c, i, 0);
+		if (status) {
+			return status;
+		}
+	}
+
+	status = is31fl3216a_write_reg(i2c, IS31FL3216A_REG_UPDATE, 0);
+	if (status) {
+		return status;
+	}
+
+	return is31fl3216a_write_reg(i2c, IS31FL3216A_REG_CONFIG, 0x00);
+
+}
+
+static int is31fl3216a_init(const struct device *dev)
+{
+	const struct is31fl3216a_cfg *config = dev->config;
+
+	LOG_DBG("Initializing @0x%x...", config->i2c.addr);
+
+	if (!device_is_ready(config->i2c.bus)) {
+		LOG_ERR("I2C device not ready");
+		return -ENODEV;
+	}
+
+	return is31fl3216a_init_registers(&config->i2c);
+}
+
+static const struct led_driver_api is31fl3216a_led_api = {
+	.set_brightness = is31fl3216a_led_set_brightness,
+	.on = is31fl3216a_led_on,
+	.off = is31fl3216a_led_off,
+};
+
```

Comment:
Note that the SSD bit could be used to implement a PM action callback. That would be nice for this driver to support the suspend and resume PM actions :)

Again this is not mandatory. This is just a suggestion that can even be added later with another PR.

Reply:
Yes makes sense but since I don't have any mean to test it for now, I will let this to the one who will need to implement it :wink:

Reply:
OK